### PR TITLE
Add options single insert

### PR DESCRIPTION
**Pull request by mtparet**
#### Description ####

Add an option to force doing a single insert so we can ensure the import will be atomic without transaction or will raise an error.

### CODE REVIEW

File: lib/activerecord-import/adapters/postgresql_adapter.rb [hunk 1]
```diff
@@ -44,7 +44,7 @@ def post_sql_statements( table_name, options ) # :nodoc:
       if (options[:ignore] || options[:on_duplicate_key_ignore]) && !options[:on_duplicate_key_update] && !options[:recursive]
         sql << sql_for_on_duplicate_key_ignore( table_name, options[:on_duplicate_key_ignore] )
       end
-    elsif options[:on_duplicate_key_ignore] && !options[:on_duplicate_key_update]
+    elsif options[:on_duplicate_key_ignore] && !options[:on_duplicate_key_update] && !options[:on_duplicate_raise]
```
Comment (jkowens) [inline lib/activerecord-import/adapters/postgresql_adapter.rb]: I don't think this does what you are thinking. This part of the conditional just prints a warning that the installed version of PostgreSQL doesn't support the option :on_duplicate_key_ignore.

File: lib/activerecord-import/adapters/mysql_adapter.rb [hunk 2]
```diff
@@ -77,7 +77,7 @@ def add_column_for_on_duplicate_key_update( column, options = {} ) # :nodoc:
       when Array then columns << column.to_sym unless columns.include?(column.to_sym)
       when Hash then columns[column.to_sym] = column.to_sym
       end
-    elsif !options[:ignore] && !options[:on_duplicate_key_ignore]
+    elsif !options[:ignore] && !options[:on_duplicate_key_ignore] && !options[:on_duplicate_raise]
```
Comment (jkowens) [inline lib/activerecord-import/adapters/mysql_adapter.rb]: To disable :on_duplicate_key_update pass the option `on_duplicate_key_update: false`

**Comment by jkowens** [comment]
I don't think there is a need for the :on_duplicate_raise option, if you update the PR I'll consider the option :single_insert. 

Can you add a test for the :single_insert option?

File: lib/activerecord-import/adapters/mysql_adapter.rb (hunk 2 above)
Comment (mtparet) [inline lib/activerecord-import/adapters/mysql_adapter.rb]: thanks, I'll give a try!

File: lib/activerecord-import/adapters/postgresql_adapter.rb (hunk 1 above)
Comment (mtparet) [inline lib/activerecord-import/adapters/postgresql_adapter.rb]: Indeed I copied the logic although here it is not relevant.

File: lib/activerecord-import/adapters/mysql_adapter.rb (hunk 2 above)
Comment (mtparet) [inline lib/activerecord-import/adapters/mysql_adapter.rb]: I just tried and it does not raise an ActiveRecord NotUnique exception on my side.. (cf my comment below about tests)

**Comment by mtparet** [comment]
Yes, I added a test for it in our project and I will put here.

File: lib/activerecord-import/adapters/mysql_adapter.rb (hunk 2 above)
Comment (mtparet) [inline lib/activerecord-import/adapters/mysql_adapter.rb]: see tests in my commit https://github.com/zdennis/activerecord-import/pull/384/commits/6c544815733a19f76bbf0b6859363f7469429519

Comment (mtparet) [inline lib/activerecord-import/adapters/mysql_adapter.rb]: My bad, I tested again and it's working as attended, thanks!

**Comment by mtparet** [comment]
I remove the :on_duplicate_raise option and I added a test for the single_insert option. I'm not sure if it possible to make it less "hacky".

**Comment by jkowens** [comment]
I'm not sure, but I think if the import is broken into multiple inserts they are wrapped in a transaction to make sure it is atomic. From your tests is that not the case?

See: https://github.com/zdennis/activerecord-import/blob/master/lib/activerecord-import/adapters/mysql_adapter.rb#L43

**Comment by mtparet** [comment]
Yes indeed this is the current and default behavior  for mysql adapter.
But we like to keep control if we want multiple insert inside a transaction (and auto-sizing) or not:

* We use https://github.com/basecamp/marginalia, to add automatic comment to mysql request, and this increase the size of the mysql request after active-import computed how much chunk it should import at once and so it go above the max_allowed_packet. ( perhaps we could have another fix/feature to support this case)
* Also we have to separate the chunck before using active-import because it enables us to not use a transaction around the import and so we are not locking the table during a long time.

**Comment by mtparet** [comment]
> From your tests is that not the case?
As it  this is a single insert there is no transaction.

**Comment by jkowens** [comment]
So after seeing what's involved, let's just forget about trying to add tests for this right now. Could you change the option name to `force_single_insert`. I think that makes the intention clear enough.

**Comment by jkowens** [comment]
@mtparet I cherry picked your commit, but changed the option to `:force_single_insert`. Thanks for your work on this one. The option is now available on `master` and I'm planning on cutting a new release here shortly.

**Comment by mtparet** [comment]
thanks a lot @jkowens !

**Comment by jkowens** [comment]
This has just been released in v0.17.2 💥

**Pull request closed without merging**